pkg/util/env/env.go

```diff
@@ -0,0 +1,138 @@
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kapi "k8s.io/kubernetes/pkg/api"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/fieldpath"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+// ResourceStore defines a new resource store data structure
+type ResourceStore struct {
+	SecretStore    map[string]*kapi.Secret
+	ConfigMapStore map[string]*kapi.ConfigMap
+}
+
+// NewResourceStore returns a pointer to a new resource store data structure
+func NewResourceStore() *ResourceStore {
+	return &ResourceStore{
+		SecretStore:    make(map[string]*kapi.Secret),
+		ConfigMapStore: make(map[string]*kapi.ConfigMap),
+	}
+}
+
+// getSecretRefValue returns the value of a secret in the supplied namespace
+func getSecretRefValue(client kclientset.Interface, namespace string, store *ResourceStore, secretSelector *kapi.SecretKeySelector) (string, error) {
+	secret, ok := store.SecretStore[secretSelector.Name]
+	if !ok {
+		var err error
+		secret, err = client.Core().Secrets(namespace).Get(secretSelector.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		store.SecretStore[secretSelector.Name] = secret
+	}
+	if data, ok := secret.Data[secretSelector.Key]; ok {
+		return string(data), nil
+	}
+	return "", fmt.Errorf("key %s not found in secret %s", secretSelector.Key, secretSelector.Name)
+
+}
+
+// getConfigMapRefValue returns the value of a configmap in the supplied namespace
+func getConfigMapRefValue(client kclientset.Interface, namespace string, store *ResourceStore, configMapSelector *kapi.ConfigMapKeySelector) (string, error) {
+	configMap, ok := store.ConfigMapStore[configMapSelector.Name]
+	if !ok {
+		var err error
+		configMap, err = client.Core().ConfigMaps(namespace).Get(configMapSelector.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+		store.ConfigMapStore[configMapSelector.Name] = configMap
+	}
+	if data, ok := configMap.Data[configMapSelector.Key]; ok {
+		return string(data), nil
+	}
+	return "", fmt.Errorf("key %s not found in config map %s", configMapSelector.Key, configMapSelector.Name)
+}
+
+// getFieldRef returns the value of the supplied path in the given object
+func getFieldRef(obj runtime.Object, from *kapi.EnvVarSource) (string, error) {
+	return fieldpath.ExtractFieldPathAsString(obj, from.FieldRef.FieldPath)
+}
+
+// getResourceFieldRef returns the value of a resource in the given container
+func getResourceFieldRef(from *kapi.EnvVarSource, c *kapi.Container) (string, error) {
+	return fieldpath.InternalExtractContainerResourceValue(from.ResourceFieldRef, c)
+}
+
+// GenEnvVarRefValue returns the value referenced by the supplied EnvVarSource given the other supplied information
+func GetEnvVarRefValue(f *clientcmd.Factory, kc kclientset.Interface, ns string, store *ResourceStore, from *kapi.EnvVarSource, obj runtime.Object, c *kapi.Container) (string, error) {
```

Comment:
Things in pkg/util are not allowed to reference clientcmd.Factory.  Only things in a pkg/cmd scope may do that.  This needs to be refactored to not reference that from here.  It's also not really appropriate to accept client interfaces into util packages unless they are much more tightly scoped.  I'd have expected this package to be `pkg/pod/envresolve` or similar.